File: backend/schemas/microlensing.py
from pydantic import BaseModel, Field


class MicrolensingPoint(BaseModel):
    hjd: float
    site: str
    magnitude: float
    mag_error: float


class MicrolensingLightCurveResponse(BaseModel):
    target_id: str
    points: list[MicrolensingPoint]
    x_label: str
    y_label: str
    extraction_mode: str = "quick"
    requested_sites: list[str] = Field(default_factory=list)
    included_sites: list[str] = Field(default_factory=list)
    missing_sites: list[str] = Field(default_factory=list)
    sampled_observation_ids: dict[str, list[str]] = Field(default_factory=dict)
    reference_observation_ids: dict[str, str] = Field(default_factory=dict)
    excluded_observation_ids: dict[str, list[str]] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    is_complete: bool = True


class MicrolensingFitInputPoint(BaseModel):
    hjd: float
    magnitude: float
    mag_error: float


class MicrolensingFitRequest(BaseModel):
    target_id: str
    points: list[MicrolensingFitInputPoint]
    t0_init: float | None = None
    u0_init: float | None = None
    tE_init: float | None = None


class MicrolensingModelPoint(BaseModel):
    hjd: float
    magnitude: float


class MicrolensingFitResponse(BaseModel):
    t0: float
    u0: float
    tE: float
    mag_base: float
    t0_err: float
    u0_err: float
    tE_err: float
    mag_base_err: float
    chi2_dof: float
    model_curve: list[MicrolensingModelPoint]


class MicrolensingPixelCoordinate(BaseModel):
    x: float
    y: float


class MicrolensingPreviewFrameMetadata(BaseModel):
    frame_index: int
    observation_id: str
    hjd: float
    site: str
    filter_band: str | None = None
    exposure_sec: float | None = None
    airmass: float | None = None
    magnitude: float
    mag_error: float
    baseline_magnitude: float
    magnification: float


class MicrolensingPreviewResponse(BaseModel):
    target_id: str
    site: str
    site_label: str
    frame_index: int
    frame_count: int
    sample_frame_indices: list[int]
    cutout_size_px: int
    cutout_width_px: int
    cutout_height_px: int
    preview_width_px: int
    preview_height_px: int
    target_position: MicrolensingPixelCoordinate
    raw_target_position: MicrolensingPixelCoordinate
    aligned_target_position: MicrolensingPixelCoordinate
    reference_target_position: MicrolensingPixelCoordinate
    reference_frame_index: int
    reference_candidate_indices: list[int] = Field(default_factory=list)
    reference_observation_id: str
    reference_hjd: float
    registration_dx_px: float
    registration_dy_px: float
    registration_quality_score: float
    registration_hit_limit: bool
    registration_warning: str | None = None
    frame_metadata: MicrolensingPreviewFrameMetadata
    raw_image_data_url: str
    aligned_image_data_url: str
    reference_image_data_url: str
    difference_image_data_url: str

File: frontend/src/types/microlensing.ts
export interface MicrolensingPoint {
  hjd: number;
  site: string;
  magnitude: number;
  mag_error: number;
}

export interface MicrolensingLightCurveResponse {
  target_id: string;
  points: MicrolensingPoint[];
  x_label: string;
  y_label: string;
  extraction_mode: string;
  requested_sites: string[];
  included_sites: string[];
  missing_sites: string[];
  sampled_observation_ids: Record<string, string[]>;
  reference_observation_ids: Record<string, string>;
  excluded_observation_ids: Record<string, string[]>;
  warnings: string[];
  is_complete: boolean;
}

export interface MicrolensingFitInputPoint {
  hjd: number;
  magnitude: number;
  mag_error: number;
}

export interface MicrolensingFitRequest {
  target_id: string;
  points: MicrolensingFitInputPoint[];
  t0_init?: number | null;
  u0_init?: number | null;
  tE_init?: number | null;
}

export interface MicrolensingModelPoint {
  hjd: number;
  magnitude: number;
}

export interface MicrolensingFitResponse {
  t0: number;
  u0: number;
  tE: number;
  mag_base: number;
  t0_err: number;
  u0_err: number;
  tE_err: number;
  mag_base_err: number;
  chi2_dof: number;
  model_curve: MicrolensingModelPoint[];
}

export interface MicrolensingPixelCoordinate {
  x: number;
  y: number;
}

export interface MicrolensingPreviewFrameMetadata {
  frame_index: number;
  observation_id: string;
  hjd: number;
  site: string;
  filter_band: string | null;
  exposure_sec: number | null;
  airmass: number | null;
  magnitude: number;
  mag_error: number;
  baseline_magnitude: number;
  magnification: number;
}

export interface MicrolensingPreviewResponse {
  target_id: string;
  site: string;
  site_label: string;
  frame_index: number;
  frame_count: number;
  sample_frame_indices: number[];
  cutout_size_px: number;
  cutout_width_px: number;
  cutout_height_px: number;
  preview_width_px: number;
  preview_height_px: number;
  target_position: MicrolensingPixelCoordinate;
  raw_target_position: MicrolensingPixelCoordinate;
  aligned_target_position: MicrolensingPixelCoordinate;
  reference_target_position: MicrolensingPixelCoordinate;
  reference_frame_index: number;
  reference_candidate_indices: number[];
  reference_observation_id: string;
  reference_hjd: number;
  registration_dx_px: number;
  registration_dy_px: number;
  registration_quality_score: number;
  registration_hit_limit: boolean;
  registration_warning: string | null;
  frame_metadata: MicrolensingPreviewFrameMetadata;
  raw_image_data_url: string;
  aligned_image_data_url: string;
  reference_image_data_url: string;
  difference_image_data_url: string;
}

export interface MicrolensingPreviewBundleResponse {
  target_id: string;
  site: string;
  focus_frame_index: number;
  reference_frame_index: number;
  bundle_frame_indices: number[];
  previews: MicrolensingPreviewResponse[];
}
